framework/src/org/checkerframework/framework/util/GenericVPUtil.java

```diff
@@ -0,0 +1,363 @@
+package org.checkerframework.framework.util;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.element.TypeParameterElement;
+import javax.lang.model.type.DeclaredType;
+import javax.lang.model.type.TypeKind;
+import org.checkerframework.framework.type.AnnotatedTypeFactory;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedArrayType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedDeclaredType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedNullType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedTypeVariable;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedWildcardType;
+import org.checkerframework.framework.type.AnnotatedTypeReplacer;
+import org.checkerframework.javacutil.Pair;
+
+/**
+ * Generic version of VPUtil. It has two major clients: one is on framework side, the other is on
+ * inference side. See {@link FrameworkVPUtil} for more information.
+ *
+ * @author tamier
+ * @author wmdietl
+ * @param <T> AnnotationMirror(framework side) or Slot(inference side)
+ */
+public abstract class GenericVPUtil<T> {
```

Comment:
@wmdietl I don't quite understand this. What kind of interface do we need? Those public methods are specific only to vputil, do they need to be included in a separate interface, do you mean?